src/app/modules/webhook/webhook.consumer.ts

```diff
@@ -0,0 +1,247 @@
+import aws from 'aws-sdk'
+import https from 'https'
+import mongoose from 'mongoose'
+import { errAsync, okAsync, ResultAsync } from 'neverthrow'
+import { Consumer } from 'sqs-consumer'
+
+import { SubmissionWebhookInfo } from '../../../types'
+import config from '../../config/config'
+import { createLoggerWithLabel } from '../../config/logger'
+import { getEncryptSubmissionModel } from '../../models/submission.server.model'
+import { transformMongoError } from '../../utils/handle-mongo-error'
+import { PossibleDatabaseError } from '../core/core.errors'
+import { SubmissionNotFoundError } from '../submission/submission.errors'
+
+import {
+  WebhookNoMoreRetriesError,
+  WebhookPushToQueueError,
+  WebhookRetriesNotEnabledError,
+  WebhookValidationError,
+} from './webhook.errors'
+import { WebhookQueueMessage } from './webhook.message'
+import { WebhookProducer } from './webhook.producer'
+import * as WebhookService from './webhook.service'
+import { isSuccessfulResponse } from './webhook.utils'
+
+const logger = createLoggerWithLabel(module)
+const EncryptSubmission = getEncryptSubmissionModel(mongoose)
+
+/**
+ * Starts polling a queue for webhook messages.
+ * @param queueUrl URL of queue from which to consume messages
+ * @param producer Producer which can be used to enqueue messages
+ */
+export const startWebhookConsumer = (
+  queueUrl: string,
+  producer: WebhookProducer,
+): void => {
+  const app = Consumer.create({
```

Comment:
Should we be specifying a duration for the `pollingWaitTimeMs` parameter to enable long polling?

Reply:
I don't think so. long polling is a feature of the queue, not of the consumer. we enable long polling on the AWS console by specifying the queue's `ReceiveMessage` wait time, so SQS knows how long to wait before sending a response back to the consumer.

on the other hand, `pollingWaitTimeMs` specifies how long the _consumer_ should wait in between each poll (see [implementation](https://github.com/bbc/sqs-consumer/blob/master/src/consumer.ts)). so e.g. if `ReceiveMessage` (on the **queue**) is 20 seconds and `pollingWaitTimeMs` (on the **consumer**) is 30s, then the following happens:
- Consumer polls the queue
- Assuming queue is empty, queue waits for 20 seconds then sends back an empty response
- Consumer waits for 30 seconds, then polls the queue again

is this consistent with your understanding?

Reply:
Oh yes that's right. Is there any reason why using the `ReceiveMessage` wait time is more preferable than the consumer's wait time? The main objective of long polling is to minimise the number of API calls to the queue right? It seems like `pollingWaitTimeMs` allows us to set an arbitrary wait time. If this is so, we can adjust according to an approximate of how crowded the queue is. I think we can get an approximation of the number of messages in the queue by using `getQueueAttributes`, which has a `ApproximateNumberOfMessages` attribute. Will this be useful?

Reply:
I think `ReceiveMessage` makes more sense because the queue is aware of whether it contains messages, whereas the consumer is not. tbh it seems to me that using a combination of `pollingWaitTimeMs` and `ApproximateNumberOfMessages` is just a workaround to achieve the same effect as `ReceiveMessageWaitTime`.

let's say the queue is empty for a long time. using `ReceiveMessageWaitTime` (let's say 10s) means that the queue will wait for 20s before responding to a poll. using `pollingWaitTimeMs` (let's say 20s) gives roughly the same effect.

but now let's say the queue is populated. `ReceiveMessageWaitTime` doesn't matter here, as the queue will respond immediately to polls as long as it contains messages. using `pollingWaitTimeMs` means the consumer will wait between polls regardless of whether the queue is populated, meaning messages may get bunched up in the queue. using `pollingWaitTimeMs` + `ApproximateNumberOfMessages` allows us to get around this, but with additional complexity as well as an additional API call per poll (to get the approx number of messages), just to achieve the same effect as `ReceiveMessageWaitTime`.

wdyt?

Reply:
Okay that's true! I was thinking more along the lines of checking the approximate number of messages every few minutes  and not before every poll. The main motivation for my suggestion is that the maximum `ReceiveMessageWaitTime` is only 20 seconds if I remember correctly, so using `pollingWaitTimeMs` can potentially help us increase the duration in between calls. But yep I agree with you too that using `ReceiveMessageWaitTime` achieves almost the same effect especially if our queues are seldom empty for long periods of time. Thanks for the discussion :)

Reply:
sure, I agree that's a good solution in case we ever need to increase the waiting time beyond 20 seconds one day!